File: java/src/main/java/org/endeavourhealth/library/model/workflow/Task.java
package org.endeavourhealth.library.model.workflow;

import lombok.Getter;
import org.endeavourhealth.library.model.tripletree.TTIriRef;
import org.endeavourhealth.library.model.workflow.task.TaskHistory;
import org.endeavourhealth.library.model.workflow.task.TaskState;
import org.endeavourhealth.library.model.workflow.task.TaskType;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Getter
public class Task {

  private TTIriRef id;
  private String createdBy;
  private TaskType type;
  private TaskState state;
  private String assignedTo;
  private LocalDateTime dateCreated;
  private List<TaskHistory> history;
  private String hostUrl;

  public Task(TTIriRef id, String createdBy, TaskType type, TaskState state, String assignedTo, LocalDateTime dateCreated, List<TaskHistory> history, String hostUrl) {
    this.id = id;
    this.createdBy = createdBy;
    this.type = type;
    this.state = state;
    this.assignedTo = assignedTo;
    this.dateCreated = dateCreated;
    this.history = history;
    this.hostUrl = hostUrl;
  }

  public Task() {
  }

  public Task setId(TTIriRef id) {
    this.id = id;
    return this;
  }

  public Task setCreatedBy(String createdBy) {
    this.createdBy = createdBy;
    return this;
  }

  public Task setType(TaskType type) {
    this.type = type;
    return this;
  }

  public Task setState(TaskState state) {
    this.state = state;
    return this;
  }

  public Task setAssignedTo(String assignedTo) {
    this.assignedTo = assignedTo;
    return this;
  }

  public Task setDateCreated(LocalDateTime dateCreated) {
    this.dateCreated = dateCreated;
    return this;
  }

  public Task setHistory(List<TaskHistory> history) {
    this.history = history;
    return this;
  }

  public Task addTaskHistory(TaskHistory taskHistory) {
    if (null == this.history) this.history = new ArrayList<>();
    this.history.add(taskHistory);
    return this;
  }

  public Task setHostUrl(String hostUrl) {
    this.hostUrl = hostUrl;
    return this;
  }
}

File: src/interfaces/Task.ts
import z from "zod";

import { TaskType } from "@/enums";
import { TaskState } from "@/enums";

import { TTIriRef, TTIriRefSchema } from "./TTIriRef";
import { TaskHistorySchema } from "./TaskHistory";

// export interface Task {
//   id?: TTIriRef;
//   createdBy?: string;
//   type?: TaskType;
//   state?: TaskState;
//   assignedTo?: string;
//   dateCreated?: Date;
//   history?: TaskHistory[];
//   hostUrl?: string;
// }

export const TaskSchema = z.strictObject({
  id: TTIriRefSchema,
  createdBy: z.string(),
  type: z.enum(TaskType),
  state: z.enum(TaskState),
  assignedTo: z.string(),
  dateCreated: z.date(),
  history: z.array(TaskHistorySchema).prefault([]),
  hostUrl: z.url()
});

export type Task = z.output<typeof TaskSchema>;
